### PR TITLE
Add transmit broadcasting with callbacks

### PR DESCRIPTION
**Pull request by mjeffrey18**
## Purpose

Bring a few features from ActionCable

## Scope

- [x] Add functionality to allow callbacks to be run after a channel has been subscribed to
- [x] Allow many after_subscribed callbacks
- [x] Allow messages to be transmitted only to the current_user for the connection subscribing
- [x] Allow mass broadcast messages to be sent to everyone subscribed to the channel
- [x] Ensure safe parsing/encoding of string messages received from Redis
- [x] Refactor server class to wrap subscribe_channel/unsubscribe_channel calls in mutex and also minor refactoring - taken from @jgaskins forks

### CODE REVIEW

**Review by jwoertink — APPROVED** [review]
I love the refactors you did in here! This is amazing 🚀  It all looks good to me 👍

**Review by russ — APPROVED** [review]
Nice! These features will be really great to use.

**Comment by mjeffrey18** [comment]
hey @russ / @jwoertink I've changed the names of the transmit and broadcast methods to be closed to the rails way, which seems to make more sense in regards to their functions also.

Let me know what you think?

**Review by jwoertink — COMMENTED** [review]
Nice. This feels pretty good. Just one typo in the README, but otherwise looks ok to me.

File: README.md [hunk 1]
```diff
@@ -121,6 +123,45 @@ class ChatChannel < ApplicationCable::Channel
...
+  def broadcast_welcome_pack_to_single_subscribed_user
+    transmit({ "welcome_pack" => "some cool stuff for this single user" })
+  end
+
+  # On the other hand,
+  # if you want to broadcast a message
+  # to all subscribers connected to this channel
+  #
+  # use -> broadcast(message)
+  def broadcast_welcome_pack_to_single_subscribed_user
```
[the remark this answers is not in the document]
Comment (mjeffrey18) [reply README.md]: good catch!

File: src/cable/channel.cr [hunk 2]
```diff
@@ -8,6 +8,19 @@ module Cable
...
+    end
+
+    # Run `block` after the subscription is created.
+    macro after_subscribed(*callbacks)
+      def run_after_subscribed_callbacks
+        {% for callback in callbacks %}
+          {{callback.id}}
+        {% end %}
+      end
+    end
```
Comment (fernandes) [inline src/cable/channel.cr:22]: I never thought too much about callbacks, I loved how simple you implemented it! that's very nice

**Comment by fernandes** [comment]
After merging master to solve `README.md` conflict one test is breaking, not sure if changes are incompatible, I can take a look in a few

**Comment by mjeffrey18** [comment]
hey, @fernandes this one is ready now. It had some rebase issues so I reset the branch and walked through it to sort out all the issues.

**Comment by fernandes** [comment]
thank you @mjeffrey18 !! here we gooo 🚀

**Pull request merged**